가위바위보 게임 [STEP 1] gogochang

gogochang opened the pull request:
@GREENOVER 
안녕하세요 그린
이번에는 TDD를 진행하면서
이전 STEP에서 배웠던 내용을 적용하였습니다.

이번 STEP에서는 진행하면서  TDD에 큰 어려움은 없었던것 같습니다. (제가 제대로 한거라면...)
TDD를 하다보면서 타입들을 좀 더 적절하게 나누게 되었습니다. 
이 과정에서 생긴 문제점에 대해 조언을 얻고싶습니다.

## 조언을 얻고 싶은 부분
### 프로퍼티의 외부접근 제어

- `HandGame`의 `windCount`,`loseCount`,`drawCount`점수는 이번 요구사항에 중요한 데이터라고 생각해서 외부에서 접근하지 않도록 구현하고 싶었습니다. 처음에 `HandGame` 프로토콜에서 가위바위보 우승자를 결정하는 `determinWinner()`의 결과로 프로토콜에 작성된 `windCount`,`loseCount`,`drawCount`를 업데이트 해주었습니다. 그렇다보니 프로토콜에서는 get, set 둘다 키워드를 작성해두어야 했기 때문에 외부에서 접근해서 변경이 가능한 상태였습니다.
- 그렇다고 `windCount`,`loseCount`,`drawCount`의 set 키워드를 제거하고 서브클래스에서 하나씩 다 구현하기에는 중복코드가 발생하였습니다. 해결 방법으로는`windCount`,`loseCount`,`drawCount`을 가지는 `Score` 구조체타입을 새로 분리하여 `HandGame` 내부에서 사용하게 수정했습니다. `Score`에 mutating이 많아지는 부분 때문에 더 좋은 방법이 없는지 조언을 얻고싶습니다.
      
#### 변경전

<img width="671" alt="2024-02-06_11 35 33" src="https://github.com/yagom-academy/refactoring-rps-game/assets/54075367/73587f7d-69ab-4c30-bf1d-a75b7706ad73">

gogochang committed:
[STEP1] 승패 판결을 위한 기능 TDD

gogochang committed:
[STEP1]
	- 삼세판을 이기면 승리하는 기능 TDD
	- 삼세판 끝나고 승패가 갈리면 초기화 하는 기능 TDD
	- 타입 분리

GREENOVER commented on RockPaperScissors/GameResult.swift:
```diff
@@ -0,0 +1,39 @@
+    case playing
+    case ready
+    
+    var result: String {
```
해당 프로퍼티를 호출하게 된다면 `gameResult.result`와 같은 네이밍처럼 지어지고 호출될것 같은데 result가 중복되서 조금 어색해보이는데 네이밍들을 바꿔보면 어떨까요~?

GREENOVER commented on RockPaperScissors/Score.swift:
```diff
@@ -0,0 +1,57 @@
+        var index = 0
+        
+        if totalScore == 0 {
+            index = GameState.GAME_READY
```
`return .win`과 같이 바로 적용시켜도 될것 같은데 이렇게 index를 만들고 한번 더 거치게 해주신 이유가 궁금합니다 🙋🏻

gogochang replied:
@GREENOVER 
이 코드 작성과정중에 
rawvalue를 통해 GameReulst를 반환하면서
rawvalue인 Int타입 매직넘버 피하기 위해 GameState를 추가하게 되었습니다.
다시 생각해보니 rawvalule사용 없이 열거형 자체의 케이스를 통해사용하기 때문에 필요없는것 같습니다.
GameState도 동일하게 의미가 없는것같습니다.

GREENOVER commented on RockPaperScissors/Score.swift:
```diff
@@ -0,0 +1,57 @@
+    var counts: [Int] { [winCount, loseCount, drawCount] }
+    var totalScore: Int { counts.reduce(0,+) }
+    
+    var correuntWinlose: String { "\(winCount)승 \(loseCount)패 \(drawCount)무" }
```
`correunt`라는 단어가 어떤 의미인지 알 수 있을까요!?

gogochang replied:
current 인데.. 오타입니다 ㅎㅎ;;

GREENOVER commented on RockPaperScissors/GameResult.swift:
```diff
@@ -0,0 +1,39 @@
+    }
+}
+
+enum GameState {
```
꼭 필요한건지 궁금해요 🤔

gogochang replied:
필요없는것같습니다!.

GREENOVER commented on RockPaperScissors/HandStrategy.swift:
```diff
@@ -0,0 +1,26 @@
+    }
+}
+
+struct ComputerHand: HandStrategy {
+    private(set) var name: String
+}
+
+struct UserHand: HandStrategy {
+    private(set) var name: String
+}
```
name으로도 컴퓨터, 나 를 표현하고 있지만 별도 두 타입으로 또 구분해주신 이유가 궁금합니다 🙋🏻
```swift
protocol HandStrategy {
    var name: String { get }
    var randomHand: String { get }
}

struct Hand: HandStrategy {
    private(set) var name: String
    
    var randomHand: String {
        return ["✊","✌️","🖐️"].randomElement() ?? ""
    }
}
```
요렇게도 하나의 방법으로 뭔가 동일 역할을 수행하는걸 단순화해볼 수 있지 않을까 생각이 들어서요!
물론, 명확하게 구분짓는것도 좋아요ㅎㅎ
스타일의 차이기에 고고창의 스타일 이유가 궁금했습니다 🙋🏻

gogochang replied:
@GREENOVER 
의도는 기능추가에 열려있는것을 고려하고자 작성했습니다.
예를 들면 예전 오락게임에서 히든캐릭터같이 특수능력을 가진 캐릭터를 고르듯이
`ComputerHand`, `UserHand`말고도 `SuperHand`라는 이름으로 프로토콜기능 이외 특수한 능력을
가지는 타입도 추가할 수 있지 않을까 생각으로 위 코드와같이 작성했습니다. 
무조건 이긴다던지, 묵만 내는 Hand같이요..ㅎㅎ;

GREENOVER commented on RockPaperScissors/RockPaperScissorsTests/RockPaperScissorsTests.swift:
```diff
@@ -0,0 +1,185 @@
+@testable import RockPaperScissors
+
+final class RockPaperScissorsTests: XCTestCase {
+    var sut: RPSGame!
```
테스트 코드긴 하지만, 강제 언래핑을 지양해보면 어떨까요~?

GREENOVER commented on RockPaperScissors/RockPaperScissorsTests/RockPaperScissorsTests.swift:
```diff
@@ -0,0 +1,185 @@
+    var sut: RPSGame!
+    
+    override func setUpWithError() throws {
+        
```
`try super.setUpWithError()`와 같이 super의 메서드 코드를 사용하지 않은 이유가 궁금합니다 🙋🏻
꼭 필요하다는것은 절대 아닙니다ㅎㅎ

gogochang replied:
@GREENOVER 
특별한 의도가 있기보다는 빠트린것 같습니다. 그렇다고 아직 명확히 그 의미를 파악하진 못했습니다.
super.setUpWithError()는 XCTestCase의 상속받아 override를 했기 때문에 
수퍼클래스의 메서드를 사용하는것 까지는 이해했지만
그 메서드의 기능이 왜 필요하고, 어떤 기능이 있는지 아직 잘 모르겠습니다.
좀 더 찾아봐야할것같습니다.

GREENOVER commented on RockPaperScissors/RockPaperScissorsTests/RockPaperScissorsTests.swift:
```diff
@@ -0,0 +1,185 @@
+    }
+    
+    //MARK: - Rock
+    func test_determineWinner에_rock_paper_입력하면_winCount가_1이다() {
```
테스트 메서드 네이밍이 한+영으로 지어질때 장/단점이 어떤거라고 생각하시나요~?

gogochang replied:
@GREENOVER 
익숙한 한글로 작성하게 되면 무슨 테스트를 하는지 쉽고 명확하게 파악이 가능합니다.
국내에서 대부분의 사람들이 한글이 익숙하기 때문에 협업에도 유리합니다.

영어로 작성하게 되면 다국적팀으로 진행하는 프로젝트에서의 협업이 유리하고,
테스트 도구를 지원하는 라이브러리나 프레임워크의 호환성이 좋습니다.

한+영으로 작성하게되면  실제 코드의 메서드나 변수는 영어로 작성하고
테스트가 무엇을 의미하는지에 대한 내용은 한글로 작성하여
한글로만 작성하는 것보다 보다 명확하게 메서드의 역할을 확인이 가능합니다.

라고 생각합니다.

GREENOVER reviewed (approved):
@gogochang 
스텝 1 수고 많으셨습니다!
특히 테스트 케이스가 정말 다양하고 거짓인 케이스들도 넣어보면서 촘촘하게 잘 구현해주신것 같아요!
요구사항을 만족하는건 더할 나위가 없었습니다ㅎㅎ
TDD를 명확하게 잘 구현해주신것 같아 크게 말씀드릴건 없는것 같습니다 👍
***
### 궁금하신 부분에 대한 저의 의견 
- 프로퍼티의 외부접근 제어
  - struct로 구현하신다는 가정하에서는 아래 방법이 시도 될 순 있을것 같아요.
    - 각 상태를 변경하는 메서드들에서 mutating을 제거하고 Score 타입으로 반환할 수도 있을것 같습니다.
    다만, 그렇게 된다면 스코어 인스턴스의 변경이 있을때마다 새 스코어 인스턴스를 생성하기에 메모리 성능에선 단점이 더 많을것 같아요.
    - private하지 않고 internal 등 접근 레벨을 변경하는것도 방법이겠지만, 캡슐화를 포기하게 됨으로 좋은 방법이라곤 생각이 크게 들진 않습니다.
  - class 타입으로 변경하는것도 고려해보면 좋지 않을까 합니다.
  적절히 사용한다면 스코어 자체의 불변성은 유지하면서 상태 변경을 해줄 수 있지 않을까 싶어요!
  
그런데 struct로 가신다고 한다면 저는 mutating 사용을 한 지금의 방식이 제일 적절해보이고 나쁘지 않다고 생각이 듭니다.
***
스텝 1 다시 한번 고생하셨습니다 😃
TDD를 꼼꼼하게 해주신게 가장 인상적이였어요.
언제든지 수정하셔도 좋고 머지 요청해주셔도 좋습니다!

gogochang committed:
[STEP1] 추가 수정
	- GameResult의 게임 결과 문구 변수명 result -> label 변경
	- GameState 제거
	- Score의 gameResult 연산 프로퍼티 수정
	- TestCode의 sut 강제언래핑 제거

gogochang commented:
@GREENOVER 
머지요청드립니다!
